---
title: "R-MarkDown-document: TNBC_TIL_analysis"
author: "Thomas Karn"
date: "May-18 2017"
output:
  word_document:
    fig_height: 3
    fig_width: 4
    fig_caption: true
    toc: yes
  pdf_document:
    fig_height: 3
    fig_width: 4
    toc: yes
  html_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



## SECTION-1  Selection of a gene expression based TNBC cohort from TCGA

We use the cgdsr package to access data from the cBIO Portal.

```{r cbioportal}
library("cgdsr")
cbiop <- CGDS("http://www.cbioportal.org/public-portal/")
# getCancerStudies(cbiop)$cancer_study_id
clidat = getClinicalData(cbiop,"brca_tcga_all")
```


### 1.1 Analysis of correlation of ESR1 gene expression from RNA-Seq and Agilent microarray platform

```{r ESR1corr}

esr1.rseq = getProfileData(cbiop,"ESR1","brca_tcga_rna_seq_v2_mrna", "brca_tcga_all")
esr1.agi = getProfileData(cbiop,"ESR1","brca_tcga_mrna", "brca_tcga_all")

# generate matrix of cases with both data for Agilent and RNA-Seq:
esr1.comp=as.data.frame(cbind(esr1.agi$ESR1, log2(esr1.rseq$ESR1+1))
        [(!is.nan(esr1.agi$ESR1)) & (!is.nan(esr1.rseq$ESR1)), ])
colnames(esr1.comp)=c("ESR1.AGI", "ESR1.RSEQ")

# correlation between Agilent and RNA-Seq:
plot(esr1.comp$ESR1.RSEQ, esr1.comp$ESR1.AGI)
cor(esr1.comp$ESR1.RSEQ, esr1.comp$ESR1.AGI)

# bimodal distribution of RNA-Seq data
hist(log2(esr1.rseq$ESR1+1), breaks=80)

```


### 1.2 Analysis of correlation of PGR gene expression from RNA-Seq and Agilent microarray platform

```{r PGRcorr}
pgr.rseq = getProfileData(cbiop,"PGR","brca_tcga_rna_seq_v2_mrna", "brca_tcga_all")
pgr.agi = getProfileData(cbiop,"PGR","brca_tcga_mrna", "brca_tcga_all")

# generate matrix of cases with both data for Agilent and RNA-Seq:
pgr.comp=as.data.frame(cbind(pgr.agi$PGR, log2(pgr.rseq$PGR+1))
          [(!is.nan(pgr.agi$PGR)) & (!is.nan(pgr.rseq$PGR)), ])
colnames(pgr.comp)=c("PGR.AGI", "PGR.RSEQ")
# correlation between Agilent and RNA-Seq:
plot(pgr.comp$PGR.RSEQ, pgr.comp$PGR.AGI)
cor(pgr.comp$PGR.RSEQ, pgr.comp$PGR.AGI)

# bimodal distribution of RNA-Seq data
hist(log2(pgr.rseq$PGR+1), breaks=80)
```


### 1.3 Analysis of correlation of HER2 gene expression from RNA-Seq and Agilent microarray platform

```{r HER2corr}
erbb2.rseq = getProfileData(cbiop,"ERBB2","brca_tcga_rna_seq_v2_mrna", "brca_tcga_all")
erbb2.agi = getProfileData(cbiop,"ERBB2","brca_tcga_mrna", "brca_tcga_all")

# generate matrix of cases with both data for Agilent and RNA-Seq:
erbb2.comp=as.data.frame(cbind(erbb2.agi$ERBB2, log2(erbb2.rseq$ERBB2+1))
            [(!is.nan(erbb2.agi$ERBB2)) & (!is.nan(erbb2.rseq$ERBB2)), ])
colnames(erbb2.comp)=c("ERBB2.AGI", "ERBB2.RSEQ")
# correlation between Agilent and RNA-Seq:
plot(erbb2.comp$ERBB2.RSEQ, erbb2.comp$ERBB2.AGI)
cor(erbb2.comp$ERBB2.RSEQ, erbb2.comp$ERBB2.AGI)

# bimodal distribution of RNA-Seq data
hist(log2(erbb2.rseq$ERBB2+1), breaks=80)
```


### 1.4 Generate TNBC dataset

```{r GenerTNBCdataset}
# Select tnbc/dnbc based on cutoffs from distribution of RNA-Seq
# define a logical selection vector
tnbc.group= !is.na(esr1.rseq) & !is.na(erbb2.rseq) &
    (log2(esr1.rseq$ESR1+1)<10) &  (log2(erbb2.rseq$ERBB2+1)<14)
colnames(tnbc.group)="tnbc"
sum(na.omit(tnbc.group))

# Generate tnbc dataset
tnbc.data= cbind(log2(esr1.rseq$ESR1+1)[tnbc.group],
                  log2(pgr.rseq$PGR+1)[tnbc.group],
                  log2(erbb2.rseq$ERBB2+1)[tnbc.group])
row.names(tnbc.data)= row.names(tnbc.group)[tnbc.group]
colnames(tnbc.data)=c("ESR1.RSEQ", "PGR.RSEQ", "ERBB2.RSEQ")

# Merge of Clinical data and tnbc dataset
# find subset in clidat corresponding to tnbc
clidat.sel=clidat[row.names(clidat)%in% row.names(tnbc.data),]
# merge tnbc.data and clinical data, left outer join:
tnbc.data= merge(tnbc.data, clidat.sel, by="row.names", all.x =TRUE)
#  "merge" creates resorted dataframe with the row.names
#      as a new first column "Row.names"
# rebuild structure (row.names):
row.names(tnbc.data)=tnbc.data$Row.names
tnbc.data=tnbc.data[,colnames(tnbc.data)!= "Row.names"]

# check residual receptor expression in tnbc dataset:
hist(tnbc.data$ESR1.RSEQ, xlim=c(0,20), breaks=40) # tnbc group
hist(log2(esr1.rseq$ESR1+1),xlim=c(0,20), breaks=80) # all samples
hist(tnbc.data$PGR.RSEQ, xlim=c(0,20), breaks=40) # tnbc group
hist(log2(pgr.rseq$PGR+1),xlim=c(0,20), breaks=80) # all samples
hist(tnbc.data$ERBB2.RSEQ, xlim=c(5,20),  breaks=40) # tnbc group
hist(log2(erbb2.rseq$ERBB2+1),xlim=c(5,20), breaks=80) # all samples
```




## SECTION-2  Gene filtering in RNA-Seq data

```{r GeneFilterRNAseq}
# Spearman correlation values between RNA-Seq and Affymetrix microarray
#  for 16,097 Jetset probes for 57 paired frozen breast cancer samples 
#  can be obtained from:
#  Suppl.Tab.S2 of Fumagalli et al. 2014, PubmedID 25412710

n208.FumagCorrel <- read.delim("2016_05_31_median_mean_n208RNASeq_vs_FumagalliCorrel.txt")

# Plot median expression vs Spearman correlation coefficient
x=n208.FumagCorrel[,c(1,3)]
plot(x)
```
```{r hexbinPlot, fig.height=5, fig.width=7}
# Use hexbin plot to display the density of the scatter
library(hexbin)
plot(hexbin(x$median, x$cor_Fumagalli, xbins=30),
     xlab="median log2 RNA-Seq expression", ylab="correlation",
     main="Correlation (RNA-Seq vs. Affy) vs. \n median RNA-Seq expression")
```
```{r MoreFumagalliPlots}
# Distribution of median expression values
hist(x$median)
# Distribution of Spearman correlation coefficients
hist(x$cor_Fumagalli)

rm(x)
```




## SECTION-3  Metagene construction

### 3.1 Metagene genes: RNA-Seq vs. Affy correlation

```{r MetageneGeneCorrel, fig.height=5, fig.width=7}
metag <- read.delim("2016_06_01_TNBC-metagenes_gene_list.txt")

par(las = 2) # labels always perpendicular to the axis
par(mar=c(7,4,4,2)+0.1) # increase bottom margin
boxplot(Correl_PMID_25412710~TNBCmetagene_RNA.Seq,
         data=metag, notch=F, col="lightblue",
         ylab="Spearman correlation",
         main="Gene correlations RNA-Seq vs Affy" )
```
```{r HistogrMetagCorrel}
par(mar=c(5.1, 4.1, 4.1, 2.1))
hist(metag$Correl_PMID_25412710)
boxplot(metag$Correl_PMID_25412710)
median(metag$Correl_PMID_25412710, na.rm=T)
summary(metag$Correl_PMID_25412710, na.rm=T)
```


### 3.2 Metagene calculation from RNA-Seq expression

```{r MetageneCalc}
# load RNAseq data of 304 genes for 208 tnbc samples
# RNAseq data of 1218 TCGA BRCA can be downloaded from UCSC Xena browser (https://tcga.xenahubs.net/download/TCGA.BRCA.sampleMap/HiSeqV2)

n304genes <- read.table("n208tnbc_n304genes_RNAseq.csv", header=TRUE, sep=";") 

# scale transposed expression data and re-transpose 
n304.expr.sca= t(scale(t(n304genes[,5:212]))) 
colnames(n304.expr.sca)=colnames(n304genes[,5:212]) 

# calculate mean expression of each metag-cluster from scaled expression for 17 metagenes 
metag17=array(NA,dim=c(0,17)) 
for (i in 1: ncol(n304.expr.sca)) { 
  mdf= as.data.frame(as.list(by(n304.expr.sca[,i], 
                                n304genes$MetagCluster17, mean))) 
  rownames(mdf)=colnames(n304.expr.sca)[i] 
  metag17=rbind(metag17, mdf) 
} 
rm(mdf) 

# merge 17 metagene expression data with tnbc.data dataframe, left outer join:
tnbc.data.meta17= merge(tnbc.data, metag17, by="row.names", all.x =TRUE) 
# "merge" command results in resorting of dataframe and loss of row.names 
#    but an additional new first column "Row.names"
# Assign new row.names from this additional column and then delete it
row.names(tnbc.data.meta17)=tnbc.data.meta17$Row.names 
tnbc.data.meta17=tnbc.data.meta17[,colnames(tnbc.data.meta17)!= "Row.names"]  

```




## SECTION-4  MATH analysis of dispersion in mutant allele frequencies

```{r MATH}
# Copy of maf file from TCGA genome.wustl.edu_BRCA.IlluminaGA_DNASeq.Level_2.1.1.0.curated.somatic.maf.txt (52MB) is available at https://portal.gdc.cancer.gov/legacy-archive/files/50d6fb1d-5bb1-4a30-9e91-6d45bd9b1c3f

# The required variant allele frequencies have been extracted in the smaller file used here: "VAF-table_genome.wustl.edu_BRCA.IlluminaGA_DNASeq.Level_2.1.1.0.curated.somatic.maf.txt""

maf.download <- read.delim(
    "VAF-table_genome.wustl.edu_BRCA.IlluminaGA_DNASeq.Level_2.1.1.0.curated.somatic.maf.txt")

all.maf = maf.download[,c("Hugo_Symbol", "Tumor_Sample_Barcode", "tumor_vaf")]

TCGA_Sample=substr(all.maf$Tumor_Sample_Barcode, 1, 15)

all.maf = cbind(TCGA_Sample, all.maf)

# calculate for each sample the median of tumor_vaf values 
med=by(all.maf$tumor_vaf, all.maf$TCGA_Sample, median)

# convert list to dataframe and transpose
med.df = t(as.data.frame(as.list(med)))
colnames(med.df)= "med.mut.AF"

# calculate MAD (Median Absolute Deviation) for each sample
MAD=by(all.maf$tumor_vaf, all.maf$TCGA_Sample, mad)

# convert list to dataframe and transpose
MAD.df= t(as.data.frame(as.list(MAD)))
colnames(MAD.df)= "MAD.mut.AF"

# calculate MATH (Mutant Allele Tumor Heterogeneity) as MATH=100*MAD/median
MATH.all =100 * MAD.df / med.df
colnames(MATH.all)= "MATH"

hist(MATH.all)

# Export MATH values:

# write.table(MATH.all, file="n982TCGA_MATH.txt",
#        row.names=TRUE, col.names = NA, quote=FALSE, sep="\t")
```




## SECTION-5  Survival analysis

```{r SurvivalAnalysis}
library("survival")

# Censor DFS at 120 months
dfs.120=tnbc.data.meta17$DFS_MONTHS
ev.120=tnbc.data.meta17$DFS_STATUS

for (i in 1:nrow(tnbc.data.meta17)) {
     if (is.na(tnbc.data.meta17$DFS_MONTHS[i]))
       {dfs.120[i]=NA ; ev.120[i]=NA}
       else 
         { if (tnbc.data.meta17$DFS_MONTHS[i] > 120)
           {dfs.120[i]=120 ; ev.120[i]="DiseaseFree"}
           else {dfs.120[i]=tnbc.data.meta17$DFS_MONTHS[i] ; ev.120=tnbc.data.meta17$DFS_STATUS}
         }
     }

# Add censored DFS to dataframe    
tnbc.data.meta17=cbind(tnbc.data.meta17, dfs.120, ev.120)

# Distributions of MHC2 metagene, B-Cell metagen, and IL8VEGF metagene
hist(tnbc.data.meta17$MHC2)
hist(tnbc.data.meta17$B.Cell)
hist(tnbc.data.meta17$IL8.VEGF)

# Since no clear bimodality observed in distributions, 
# we stay with previously established cutoffs for metagenes/signatures:
# MHC2 metagene: Upper quartile (Rody 2009, PMID 19272155)
# B-Cell metagene: Lower quartile (Rody 2011, PMID 21978456)
# IL8.VEGF metagene: Median split (Rody 2011, PMID 21978456)

```



### 5.1 MHC2/IL8VEGF signature
```{r Survival MHC2/IL8VEGF signature}


# Define upper quartile MHC2 metagene (based on Rody 2009, PMID 19272155)
MHC2.q4=tnbc.data.meta17$MHC2 > quantile(tnbc.data.meta17$MHC2, probs=0.75)
# Define below median IL8.VEGF metagene (cutoff from Rody 2011, PMID 21978456)
IL8.VEGF.q12=tnbc.data.meta17$IL8.VEGF < quantile(tnbc.data.meta17$IL8.VEGF, probs=0.5)
# Define prognostic signature
MHC2.IL8.VEGF.sig = MHC2.q4 & IL8.VEGF.q12


## Check MHC2.IL8.VEGF.sig in Survival analysis 
time=tnbc.data.meta17$dfs.120
censor= (tnbc.data.meta17$ev.120 =="Recurred/Progressed") 
strata= MHC2.IL8.VEGF.sig 
test=survfit(Surv(time, censor)~strata,conf.type="none") 
summary(test) 
plot(test, lty=c(1,3), xlab="Time", ylab="Survival Probability") 
legend(10, 0.4, c("Poor", "Good") , lty=c(1,2))


```

### 5.2 B-Cell/IL8VEGF signature
```{r Survival B-Cell/IL8VEGF signature}

# Define B-Cell metagene above lowest quartile (cutoff from Rody 2011, PMID 21978456)
B.Cell.q234=tnbc.data.meta17$B.Cell > quantile(tnbc.data.meta17$B.Cell, probs=0.25)
# Define below median IL8.VEGF metagene (cutoff from Rody 2011, PMID 21978456)
IL8.VEGF.q12=tnbc.data.meta17$IL8.VEGF < quantile(tnbc.data.meta17$IL8.VEGF, probs=0.5)
# Define prognostic signature
B.Cell.IL8.VEGF.sig = B.Cell.q234 & IL8.VEGF.q12


## Check B.Cell.IL8.VEGF.sig in Survival analysis 
time=tnbc.data.meta17$dfs.120
censor= (tnbc.data.meta17$ev.120 =="Recurred/Progressed") 
strata= B.Cell.IL8.VEGF.sig 
test=survfit(Surv(time, censor)~strata,conf.type="none") 
summary(test) 
plot(test, lty=c(1,3), xlab="Time", ylab="Survival Probability") 
legend(10, 0.4, c("Poor", "Good") , lty=c(1,2))



```


```{r sessionInfo}
dir()
sessionInfo()
```

